llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -486,6 +560,25 @@ PreservedAnalyses SpecConstantsPass::run(Module &M,
       StringRef SymID = getStringLiteralArg(CI, NameArgNo, DelInsts);
       Value *Replacement = nullptr;
 
+      Constant *DefaultValue = nullptr;
+      if (Is2020Intrinsic) {
+        // For SYCL 2020, there is a mechanism to specify the default value.
+        // It is stored as an initializer of a global variable referenced by
+        // the second argument of the intrinsic.
+        auto *GV = dyn_cast<GlobalVariable>(
+            CI->getArgOperand(NameArgNo + 1)->stripPointerCasts());
+        if (GV) {
+          assert(GV->hasInitializer() && "expected initializer");
+          auto *Initializer = GV->getInitializer();
+          assert((isa<ConstantAggregate>(Initializer) ||
+                  Initializer->isZeroValue()) &&
+                 "expected specialization_id instance");
+          // specialization_id structure contains a single field which is the
+          // default value of corresponding specialization constant.
+          DefaultValue = Initializer->getAggregateElement(0u);
+        }
+      }
```

Comment:
what GV == 0 means? please add a comment or handle

Reply:
We can meet nullptr instead of value here - this check is still needed. I left a comment above the if-.